File: csharp/protein-translation/ProteinTranslation.cs
using System;
using System.Collections.Generic;

public static class ProteinTranslation
{
    public static string[] Proteins(string rna)
    {
        var codonMap = new Dictionary<string, string>
        {
            { "AUG", "Methionine" },
            { "UUU", "Phenylalanine" }, { "UUC", "Phenylalanine" },
            { "UUA", "Leucine" }, { "UUG", "Leucine" },
            { "UCU", "Serine" }, { "UCC", "Serine" }, { "UCA", "Serine" }, { "UCG", "Serine" },
            { "UAU", "Tyrosine" }, { "UAC", "Tyrosine" },
            { "UGU", "Cysteine" }, { "UGC", "Cysteine" },
            { "UGG", "Tryptophan" },
            { "UAA", "STOP" }, { "UAG", "STOP" }, { "UGA", "STOP" }
        };

        var proteins = new List<string>();
        
        // Process RNA sequence in chunks of 3 (codons)
        for (int i = 0; i < rna.Length; i += 3)
        {
            // Ensure there are enough nucleotides left to form a codon
            if (i + 3 > rna.Length) break;

            string codon = rna.Substring(i, 3);

            // If the codon is a STOP codon, terminate translation
            if (codonMap[codon] == "STOP")
                break;

            // If the codon maps to a protein, add it to the list
            if (codonMap.ContainsKey(codon))
                proteins.Add(codonMap[codon]);
        }

        return proteins.ToArray();
    }
}

File: fsharp/protein-translation/ProteinTranslation.fs
module ProteinTranslation

  let codonToProtein = 
    dict [
      "AUG", "Methionine"
      "UUU", "Phenylalanine"; "UUC", "Phenylalanine"
      "UUA", "Leucine"; "UUG", "Leucine"
      "UCU", "Serine"; "UCC", "Serine"; "UCA", "Serine"; "UCG", "Serine"
      "UAU", "Tyrosine"; "UAC", "Tyrosine"
      "UGU", "Cysteine"; "UGC", "Cysteine"
      "UGG", "Tryptophan"
      "UAA", "STOP"; "UAG", "STOP"; "UGA", "STOP"
    ]


  let proteins rna =
        rna
        |> Seq.chunkBySize 3
        |> Seq.map (fun codon -> 
            let key = System.String codon
            match codonToProtein.TryGetValue(key) with
            | true, value -> Some value
            | false, _ -> None)
        |> Seq.takeWhile (fun protein -> protein <> Some "STOP")
        |> Seq.choose id
        |> Seq.toList
